### PR TITLE
Fix auto-splice for tibbles

### LINKED ISSUE #2326: Automatically unpack unnamed df-cols
Currently `mutate()` and `summarise()`  only work with vectorised functions: functions that take a vector as input and return a vector (or "scalar") as output. I don't see any reason why `summarise()` and `mutate()` couldn't also accept tibbles. The existing restrictions would continue to apply so that in `summarise()` the tibble would have to have exactly one row, and in `mutate()` it would have to have either one row or n rows.

In other words, the following two lines of code should be equivalent:

```R
df %>%
  summarise(mean = mean(x), sd = sd(x))

df %>%
   summarise(tibble(mean = mean(x), sd = sd(x))
```

This would allow you to extract that repeated pattern out into a function:

```R
# and hence
mean_sd <- function(df, var) {
  tibble(mean = mean(df[[var]]), sd = sd(df[[var]]))
}
df %>% 
  summarise(mean_sd(df, "x"))
```

We'd need to work on documentation to help people develop effective functions of this nature  develop tools so that you could easily specify input variables (using whatever the next iteration of lazyeval provides) and name the outputs. But that's largely a second-order concern: we can figure out those details later.

Supporting tibbles in this way would be particular useful for dplyr as it would help to clarify the nature of functions like `separate()` and `unite()` which are currently data frame wrappers around simple vector functions.

These ideas are most important for `summarise()` and `mutate()` but I think we should apply the same principles to `filter()` and `arrange()` as well.

cc @lionel- @jennybc @krlmlr

### PR DESCRIPTION
**Pull request by krlmlr**
First commit fixes tests that failed previously.

Closes #2326.

### CODE REVIEW

**Comment by krlmlr** [comment]
`pluck()` seems better here. Also, please consider the tests and the `force()`.

File: R/tbl-df.r [hunk 1]
```diff
@@ -330,7 +330,7 @@ mutate.tbl_df <- function(.data, ...) {
 
       # remember each result separately
       map2(seq_along(result), names(result), function(i, nm) {
-        mask$add(nm, map(chunks, function(.x) .x[[i]]))
+        mask$add(nm, pluck(chunks, i))
```
Comment (romainfrancois) [inline R/tbl-df.r]: nice I didn't realize compat had `pluck()`

File: R/tbl-df.r [hunk 2]
```diff
@@ -417,6 +417,7 @@ DataMask <- R6Class("DataMask",
     },
 
     add = function(name, chunks) {
+      force(chunks)
```
Comment (romainfrancois) [inline R/tbl-df.r]: Why is this needed ?

**Comment by krlmlr** [comment]
I tend to force everything that is used in a closure to avoid nasty surprises later on. The force() reveals the error seen here much earlier.

**Pull request merged**